multistream-2.0/spec.md

```diff
@@ -0,0 +1,403 @@
+useful (a) in packet-based protocols (without sessions) and (b) when initially
+negotiating a transport session (before protocols have been advertised and the
+stream multiplexer has been configured).
+
+1. `serial-stream`: A simple stream "multiplexer" that can multiplex multiple
+   streams *in serial* over the same connection. That is, it allows us to
+   negotiate a protocol, use it, and then return to multistream. It also allows
+   us to speculatively choose a single protocol and then drop back down to
+   multistream if that doesn't work.
+2. `speculative-stream`: A speculative stream "multiplexer" where the initiator
+   can speculatively initiate multiple streams and the receiver must select at
+   most one and discard the others.
+   
+All peers *must* implement `multistream/use` and *should* implement
+`serial-stream`. This combination will allow us to apply a series of quick
+connection upgrades (e.g., to multistream 3.0) with no round trips and no funny
+business (learn from past mistakes).
+
+Notes:
+
+1. The "ls" feature of multistream has been removed. While useful, this really
+   should be a *protocol*. Given the `serial-stream` protocol, this shouldn't be
+   an issue as we can run as many sub-protocols over the same stream as we want.
+2. To reduce RTTs, all protocols are unidirectional.
+3. These protocols were *also* designed to eventually support packet protocols
+   (the other reason to be unidirectional and a strong motivator for the
+   `speculative-stream` and `serial-stream` protocols).
+
+### Multistream Advertise
+
+Unspeced (for now). Really, we just need to send a mapping of protocol
+names/codecs to contextual IDs (and may be some service discovery information).
+Basically, identify.
+
+### Multistream Use
+
+The `multistream/use` protocol is simply two varint multicodecs: the
+multistream-use multicodec followed by the multicodec for the protocol to be
+used. This protocol is *unidirectional*. If the stream is bidirectional, the
+receiver must acknowledge a successful protocol negotiation by responding with
+the same multistream-use protocol sequence.
+
+Every stream starts with multistream-use. Every other protocol defined here will
+be assigned a multicodec and selected with `multistream/use.`
+
+This protocol should *also* be trivial to optimize in hardware simply by prefix
+matching (i.e., matching on the first N (usually 16-32) bits of the
+stream/message).
+
+* [ ] Q: Technically, the first multicodec is redundant. However, it acts as a
+      magic byte that allows us to figure out what's going on. Should we keep
+      it? We could just start all streams with a single multicodec representing
+      the protocol
+* [ ] Q: Should we somehow distinguish between initiator and receiver? Should we
+      distinguish between bidirectional and unidirectional? We could even bit
+      pack these options into a single byte and use this instead of the leading
+      multicodec...
+
+### Multistream Dynamic
```

Comment:
Do we really have a use case for this? Or wouldn't it be easier for peers to first announce a new code point, and then use it with `multistream/use`?

Reply:
Privacy, flexibility, feature parity with multistream 1.0. For example, I might listen on a non-enumerable set of protocols.

Concretely: We can use multistream to connect to peers within peers by connecting to the multistream endpoint `/p2p/InnerPeerId`.

> Or wouldn't it be easier for peers to first announce a new code point, and then use it with multistream/use

Note: multistream/use is for multicodecs. For custom protocols, we'd use multistream/contextual.